### PR TITLE
added exponential backoff and retries while preserving legacy Registe…

### PR DESCRIPTION
**Pull request by dancompton**
…r() call.  Added tests

### CODE REVIEW

File: portmapper.go [hunk 1]
```diff
@@ -93,49 +86,79 @@ func Unregister(name string, port int) error {
...
+func RegisterService(name string, port int, retries int) error {
 	svc := &Service{name, port, os.Getenv("HOSTNAME")}
-	ServiceMap[name] = &ServiceRegistration{Service: svc, Timestamp: 0, err: nil}
 
-	//XXX I assume that this will not exit ever
-	once.Do(RegisterServices)
-	return nil
+	var regerr error
+
+	for try := 1; try < retries; try++ {
```
Comment (grepory) [inline portmapper.go]: You should differentiate between transient and permanent failures. E.g. if the service fails validation, fail fast without retrying. Same with the Marshal() call. 

The only thing that should retry is the interaction with Etcd.

File: portmapper_test.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package portmapper
+
+import (
+	"testing"
+)
+
+func Test_RegisterService(t *testing.T) {
+	if Register("ServiceA", 4000) != nil {
+		t.Error("error registering service")
```
Comment (grepory) [inline portmapper_test.go:9]: This fails without etcd running, yes?

Comment (dancompton) [inline portmapper_test.go:9]: It does.

File: portmapper.go (hunk 1 above)
Comment (dancompton) [inline portmapper.go]: Oh, I get it now :package:

**Comment by grepory** [comment]
You can write a test that uses the library. Stand up etcd in a container in circle. Point yourself at that etcd, and test against it.

**Comment by dancompton** [comment]
Ah, that makes sense.  I need to set it up in circle.  Can we merge as is so that we can move towards bastion portmapper update?

**Comment by dancompton** [comment]
actually, I'd be more comfortable after testing again

**Comment by grepory** [comment]
RegisterService doesn't work how you think it does.

Your goroutine will return immediately and regerror will always be nil. This is because you wait on your waitgroup inside the goroutine.

Rather than do any asynchronous work here, make RegisterService() synchronous.

**Comment by grepory** [comment]
I'd like to see CI for the portmapper library added before we merge.

**Comment by dancompton** [comment]
I'm confused, I call wg.Wait() _outside_ of the goroutine.

**Comment by grepory** [comment]
Oh you're right. Sorry, my bad. There are a lot of curly braces. :curly_loop: 

Is there an argument for using goroutines at all?

Anything that isn't communicating with etcd should also come out of the for loop.

**Comment by dancompton** [comment]
It seems like I had a reason, but I can't remember it now.

**Comment by dancompton** [comment]
regardless, this will have to wait because Corey needs a bastion

**Comment by grepory** [comment]
What will have to wait?

**Pull request closed without merging**